### DISCUSSION TITLE
Image Comparison in C

### DISCUSSION THREAD

**Discussion by sethify** [libvips/libvips#2118] [Q&A]
Hello,

We have an existing image transform pipeline that crops, resizes and converts images as they are requested. We are experimenting with using libvips for this process (so far libvips has been amazing). 

As part of our the test suite we would like to compare images generated by the libvips solution vs the old solution. We don't expect them to be identical, but would like to know how different they are. Google led me here: 

https://stackoverflow.com/questions/36524206/is-possible-to-detect-if-two-images-are-the-same-using-ruby-vips8 

This solution seemed seemed very interesting. Fortunately/unfortunately we are using C (we wrote our own Go binding) and this magic doesn't appear to be as easy. 

I expect the team will have many ideas about how to perform diffing, so, access raw pixels is probably what I need. What is the best way with the C API? (Images are already in memory, loaded from a buffer).

Thank you.

**Comment by jcupitt** [comment]
Hello @sethify,

It sounds like something along the lines of spatial CIELAB would work for this case:

https://web.stanford.edu/group/vista/cgi-bin/wiki/index.php/SCIELAB

https://www.researchgate.net/publication/2629449_Color_Image_Quality_Metric_S-CIELAB_and_Its_Application_on_Halftone_Texture_Visibility

It's really simple -- just do a gauss blur before calculating delta E.

In ruby (for example):

```
def scielab(a, b, sigma)
    a.gaussblur(sigma).dE76(b.gaussblur(sigma)).avg
end
```

ie. blur both, calculate the delta E, then take the average. `sigma` is a parameter that sets the "scale" of the test, ie. how close the observer is.

**Reply by sethify** [reply]
Wow, thank you. That is very interesting. How does one get the `avg` in C?

**Reply by sethify** [reply]
I think I just want to read R,G,B and A values from the image buffer. How does one access the pixels directly in C? Thank you.

**Reply by sethify** [reply]
Would regions (as in `vips_region_new` be my ticket?

**Reply by jcupitt** [reply]
You can write it in C with eg.:

```C
    VipsImage **t = (VipsImage **)
                vips_object_local_array( object, 3 );
    double average;

    if (vips_gaussblur(in1, &t[0], sigma, NULL) ||
        vips_gaussblur(in2, &t[1], sigma, NULL) ||
        vips_dE76(t[0], t[1], &t[2], NULL) ||
        vips_avg(t[2], &average, NULL))
        return -1;
```

**Reply by sethify** [reply]
Nice. Thank you. I understand that API better now, just from this one example. This example also lead me to learn more about what Delta E is, very helpful :) Is the `vips_de00` the 2000 version?

**Reply by jcupitt** [reply]
Yes, that's right, cielab 2000. But everyone still uses 76 because it's only slightly worse, and a lot simpler.

**Reply by sethify** [reply]
I have DeltaE working, and am also trying to get an average pixel value difference as a second measure. I think this does exactly that. I've compared the output of the source image to subsequently lower quality jpegs and the value increases as predicted. My thinking is that this is the average difference of pixels RGB (and maybe A) values: 
```C
bool lv_diff(VipsImage *leftImage, VipsImage *rightImage, double *avg) 
{
    VipsObject *context = VIPS_OBJECT(vips_image_new());
    VipsImage **t = (VipsImage **)vips_object_local_array( context, 4 );
    vips_copy(leftImage, &t[0], NULL);
    vips_copy(rightImage, &t[1], NULL);

    bool error = 
        vips_subtract(t[0], t[1], &t[2], NULL) ||
        vips_abs(t[2], &t[3], NULL) ||
        vips_avg(t[3], avg, NULL);

    g_object_unref( context );
    return !error;
}
```

**Reply by sethify** [reply]
Here is the output comparing an image saved at full quality with to the same image saved at various quality levels, I think this works:

```
q=100 dE=0.0000 diff=0.0000
q=95  dE=0.1147 diff=0.9824
q=80  dE=0.2727 diff=1.7054
q=60  dE=0.3122 diff=2.0907
q=40  dE=0.4906 diff=2.4629
q=20  dE=0.8132 diff=3.2432
```

**Reply by jcupitt** [reply]
You don't need the copies, but yes, looks reasonable. dE should perform better for photographic images. I'd add a blur as well to remove differences due to small errors in positioning.

**Reply by sethify** [reply]
Thank you very much. I expect we will add more comparisons such as this: https://github.com/cloudinary/ssimulacra in an effort to quantify quality vs an ideal image.

**Reply by jcupitt** [reply]
Good idea, DSSIM is excellent.
